### PR TITLE
[Group 2] Enable nullable annotations for `Microsoft.Extensions.Caching.Abstractions`

### LINKED ISSUE #52148: Update IDistributedCache to account for nullability
## Background and Motivation

The `Get` and `GetAsync` methods of the `IDistributedCache` interface do not match the intended nullability in terms of return type. As per the API documentation:

```c#
/// <returns>The located value or null.</returns>
```

I noticed this when we were working on adding community contributed APIs, see: https://github.com/dotnet/aspnetcore/pull/32266#discussion_r622870585. 

The request here is to update the API in the `IDistributedCache` interface.

## Proposed API

```diff
namespace Microsoft.Extensions.Caching.Distributed
{
  public interface IDistributedCache
  {
-    byte[] Get(string key);
+    byte[]? Get(string key);
-    Task<byte[]> GetAsync(string key, CancellationToken token = default(CancellationToken));
+    Task<byte[]?> GetAsync(string key, CancellationToken token = default(CancellationToken));
  }
```

## Usage Examples

The return type will be update:

```diff
- byte[] data = cache.Get(key);
+ byte[]? data = cache.Get(key);
```

## Alternative Designs

The alternative is to keep the current API. This presents a difficulty when adding nullability checks in implementations of `IDistributedCache` see https://github.com/dotnet/aspnetcore/pull/32266. 

## Risks

Implementations of the `IDistributedCache` interface will need to be updated to account for the nullability changes proposed here.

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Related to #43605
Fixes: #52148

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/src/PostEvictionCallbackRegistration.cs [hunk 1]
```diff
@@ -5,8 +5,8 @@ namespace Microsoft.Extensions.Caching.Memory
 {
     public class PostEvictionCallbackRegistration
     {
-        public PostEvictionDelegate EvictionCallback { get; set; }
+        public PostEvictionDelegate EvictionCallback { get; set; } = null!;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Caching.Abstractions/src/PostEvictionCallbackRegistration.cs:8]: This is a lie. The property should be annotated as nullable.

```suggestion
        public PostEvictionDelegate? EvictionCallback { get; set; }
```

**Review by eerhardt — COMMENTED** [review]
This looks pretty good. I just have one comment.

BTW - have you locally tried building/annotating MS.Ext.Caching.Memory to ensure these annotations are correct?

**Comment by maxkoshevoi** [comment]
> have you locally tried building/annotating MS.Ext.Caching.Memory to ensure these annotations are correct?

Yep, I've also annotated `Caching.Memory`. Didn't create a PR for it, since PR for `Extensions.Options` is not merged yet, so some annotations might be wrong.

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/src/PostEvictionCallbackRegistration.cs (hunk 1 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Caching.Abstractions/src/PostEvictionCallbackRegistration.cs:8]: `CacheEntryExtensions.RegisterPostEvictionCallback` would throw on null `EvictionCallback`
Called here: https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Caching.Abstractions/src/CacheEntryExtensions.cs#L189

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Caching.Abstractions/src/PostEvictionCallbackRegistration.cs:8]: This class has a default ctor. So just doing:

```C#
new PostEvictionCallbackRegistration()
```

creates an object with a null EvictionCallback.

> CacheEntryExtensions.RegisterPostEvictionCallback would throw on null EvictionCallback

That's a bug.

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Caching.Abstractions/src/PostEvictionCallbackRegistration.cs]: Created an issue: https://github.com/dotnet/runtime/issues/64147

**Review by eerhardt — APPROVED** [review]
Thanks for this contribution, @maxkoshevoi! LGTM.

Sorry for the delay in reviewing.

**Pull request merged**